sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala

```diff
@@ -900,13 +904,80 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
               s"${SQLConf.HIVE_METASTORE_PARTITION_PRUNING_FALLBACK_ON_EXCEPTION.key} " +
               " to false and let the query fail instead.", ex)
             // HiveShim clients are expected to handle a superset of the requested partitions
-            getAllPartitionsMethod.invoke(hive, table).asInstanceOf[JSet[Partition]]
+            prunePartitionsEvalClientSide(hive, table, catalogTable, predicates)
           case ex: InvocationTargetException if ex.getCause.isInstanceOf[MetaException] =>
             throw QueryExecutionErrors.getPartitionMetadataByFilterError(ex)
         }
       }
 
-    partitions.asScala.toSeq
+    partitions.toSeq
+  }
+
+  def prunePartitionsEvalClientSide(hive: Hive,
+      table: Table,
+      catalogTable: CatalogTable,
+      predicates: Seq[Expression]): Seq[Partition] = {
+
+    val timeZoneId = SQLConf.get.sessionLocalTimeZone
+
+    // Because there is no way to know whether the partition properties has timeZone,
+    // client-side filtering cannot be used with TimeZoneAwareExpression.
+    def hasTimeZoneAwareExpression(e: Expression): Boolean = {
+      e.collectFirst {
+        case t: TimeZoneAwareExpression => t
+      }.isDefined
+    }
+
+    if (!SQLConf.get.metastorePartitionPruningEvalClientSide ||
+      predicates.isEmpty ||
+      predicates.exists(hasTimeZoneAwareExpression)) {
+      getAllPartitionsMethod.invoke(hive, table).asInstanceOf[JSet[Partition]].asScala.toSeq
+    } else {
+      try {
+        val partitionSchema = CharVarcharUtils.replaceCharVarcharWithStringInSchema(
+          catalogTable.partitionSchema)
+        val partitionColumnNames = catalogTable.partitionColumnNames.toSet
+
+        val nonPartitionPruningPredicates = predicates.filterNot {
+          _.references.map(_.name).toSet.subsetOf(partitionColumnNames)
+        }
+        if (nonPartitionPruningPredicates.nonEmpty) {
+          throw QueryCompilationErrors.nonPartitionPruningPredicatesNotExpectedError(
+            nonPartitionPruningPredicates)
+        }
+
+        val boundPredicate =
+          Predicate.createInterpreted(predicates.reduce(And).transform {
+            case att: AttributeReference =>
+              val index = partitionSchema.indexWhere(_.name == att.name)
+              BoundReference(index, partitionSchema(index).dataType, nullable = true)
+          })
+
+        def toRow(spec: TablePartitionSpec): InternalRow = {
+          InternalRow.fromSeq(partitionSchema.map { field =>
+            val partValue = if (spec(field.name) == ExternalCatalogUtils.DEFAULT_PARTITION_NAME) {
+              null
+            } else {
+              spec(field.name)
+            }
+            Cast(Literal(partValue), field.dataType, Option(timeZoneId)).eval()
+          })
+        }
+
+        val allPartitionNames = hive.getPartitionNames(
```

Comment:
This is a good idea!

Also cc @sunchao

Reply:
+1. This looks like a nice optimization.

Reply:
Looks reasonable to me as well. Just for my understanding, the rationale here is `getPartitionNames` is much cheaper in terms of CPU cost on Metastore server side, compared to fetch all `Partition` objects right?

Reply:
`getPartitionNames` 80,000 partitions are within 1s, `getAllPartitionsOf` is about 4-5 minutes.
The `Partition` object returned by `getAllPartitionsOf` contains detailed information.
This can also save the memory footprint of the driver.